File: src/TspJulia/SimulatedAnnealing.jl
module SimulatedAnnealing

export AbstractSimulatedAnnealing, SimulatedAnnealingTSP, run_sa, SAStatsProxy

using Random
using ..Elements
using ..Kernels
using ..ScheduleFns

abstract type AbstractSimulatedAnnealing end

get_cities(sa::AbstractSimulatedAnnealing)::Elements.Cities =
    throw(NotImplementedError("get_cities not implemented for $(typeof(sa))"))
get_kernel(sa::AbstractSimulatedAnnealing)::Elements.AbstractKernel =
    throw(NotImplementedError("get_kernel not implemented for $(typeof(sa))"))
get_n_iter(sa::AbstractSimulatedAnnealing)::Int =
    throw(NotImplementedError("get_n_iter not implemented for $(typeof(sa))"))
get_cooling_schedule(sa::AbstractSimulatedAnnealing)::Function =
    throw(NotImplementedError("get_cooling_schedule not implemented for $(typeof(sa))"))
get_early_stop(sa::AbstractSimulatedAnnealing)::Bool =
    throw(NotImplementedError("get_early_stop not implemented for $(typeof(sa))"))
get_stop_after(sa::AbstractSimulatedAnnealing)::Int =
    throw(NotImplementedError("get_stop_after not implemented for $(typeof(sa))"))

initial_route(sa::AbstractSimulatedAnnealing)::Elements.Route =
    Elements.Route(1:length(get_cities(sa)))

sample_from_kernel(
    sa::AbstractSimulatedAnnealing,
    route::Elements.Route,
)::Elements.Route = sample(get_kernel(sa), route)

compute_value(sa::AbstractSimulatedAnnealing, route::Elements.Route)::Float64 =
    total_distance(get_cities(sa), route)

compute_log_acceptance_prob(
    sa::AbstractSimulatedAnnealing,
    route::Elements.Route,
    new_route::Elements.Route,
    k::Int,
)::Float64 =
    (-compute_value(sa, new_route) + compute_value(sa, route)) /
    (get_cooling_schedule(sa)(k) + eps(Float64))

select_route(
    sa::AbstractSimulatedAnnealing,
    route::Elements.Route,
    new_route::Elements.Route,
    log_accept_prob::Float64,
)::Elements.Route =
    log_accept_prob > 0 || rand() < exp(log_accept_prob) ? new_route : route

stop_criteria(sa::AbstractSimulatedAnnealing, k::Int)::Bool =
    get_early_stop(sa) && k > get_stop_after(sa)

function run_sa(sa::AbstractSimulatedAnnealing)::Elements.Route
    route = initial_route(sa)
    route_value = compute_value(sa, route)
    best_route = route
    best_value = route_value
    for k ∈ 1:get_n_iter(sa)
        # route = step_sa(sa, route, k)
        new_route = sample_from_kernel(sa, route)
        log_accept_prob = compute_log_acceptance_prob(sa, route, new_route, k)
        route = select_route(sa, route, new_route, log_accept_prob)
        route_value = compute_value(sa, route)
        if route_value < best_value
            best_route = route
            best_value = route_value
        end
        if stop_criteria(sa, k)
            break
        end
    end
    return best_route
end

struct SimulatedAnnealingTSP{K <: Kernels.AbstractKernel} <: AbstractSimulatedAnnealing
    cities::Elements.Cities
    kernel::K
    n_iter::Int
    cooling_schedule::Function
    early_stop::Bool
    stop_after::Int

    function SimulatedAnnealingTSP(
        cities::Elements.Cities,
        kernel::K,
        n_iter::Int = 1_000_000,
        cooling_schedule::Function = ScheduleFns.exponential_cooling_schedule(
            T₀ = 100,
            ρ = 0.99,
        ),
        early_stop::Bool = true,
        stop_after::Int = 10_000,
    ) where {K <: AbstractKernel}
        return new{K}(cities, kernel, n_iter, cooling_schedule, early_stop, stop_after)
    end
end

get_cities(sa::SimulatedAnnealingTSP)::Elements.Cities = sa.cities
get_kernel(sa::SimulatedAnnealingTSP)::Kernels.AbstractKernel = sa.kernel
get_n_iter(sa::SimulatedAnnealingTSP)::Int = sa.n_iter
get_cooling_schedule(sa::SimulatedAnnealingTSP)::Function = sa.cooling_schedule
get_early_stop(sa::SimulatedAnnealingTSP)::Bool = sa.early_stop
get_stop_after(sa::SimulatedAnnealingTSP)::Int = sa.stop_after

mutable struct SAStatsProxy <: AbstractSimulatedAnnealing
    sa::SimulatedAnnealingTSP
    log_acceptance_probs::Vector{Float64}
    n_accept::Int64
    acceptance_ratio::Vector{Float64}
    values::Vector{Float16}

    SAStatsProxy(sa::SimulatedAnnealingTSP) = new(sa, [], 0, [], [])
end

get_cities(sa::SAStatsProxy)::Elements.Cities = sa.sa.cities
get_kernel(sa::SAStatsProxy)::Kernels.AbstractKernel = sa.sa.kernel
get_n_iter(sa::SAStatsProxy)::Int = sa.sa.n_iter
get_cooling_schedule(sa::SAStatsProxy)::Function = sa.sa.cooling_schedule
get_early_stop(sa::SAStatsProxy)::Bool = sa.sa.early_stop
get_stop_after(sa::SAStatsProxy)::Int = sa.sa.stop_after

function Base.show(io::IO, sa::SAStatsProxy)
    repr = "SAStatsProxy(log_acceptance_probs=$(length(sa.log_acceptance_probs)), n_accept=$(sa.n_accept), acceptance_ratio=$(length(sa.acceptance_ratio)), values=$(length(sa.values)))"
    return print(io, repr)
end

function sample_from_kernel(sa::SAStatsProxy, route::Elements.Route)::Elements.Route
    new_route = sample_from_kernel(sa.sa, route)
    push!(sa.values, compute_value(sa, new_route))
    return new_route
end

function compute_log_acceptance_prob(
    sa::SAStatsProxy,
    route::Elements.Route,
    new_route::Elements.Route,
    k::Int,
)::Float64
    log_accept_prob = compute_log_acceptance_prob(sa.sa, route, new_route, k)
    push!(sa.log_acceptance_probs, log_accept_prob)
    return log_accept_prob
end

function select_route(
    sa::SAStatsProxy,
    route::Elements.Route,
    new_route::Elements.Route,
    log_accept_prob::Float64,
)::Elements.Route
    if log_accept_prob > 0 || rand() < exp(log_accept_prob)
        sa.n_accept += 1
        return new_route
    else
        return route
    end
end

end  # module SimulatedAnnealing

File: src/tsp/py/simulated_annealing.py

import abc
import typing as t
from dataclasses import dataclass, field

import numpy as np
from numpy.random import Generator

from ..plotting import SAStatsProtocol
from ..utils import _EPS, cooling_schedule_function, exponential_cooling_schedule
from .elements import Cities, Route
from .kernels import Kernel, SwapKernel

__all__ = ["SimulatedAnnealingTSP", "SAStatsProxy"]


@dataclass
class AbstractSimulatedAnnealing(abc.ABC):
    """
    Abstract base class for Simulated Annealing.
    """
    
    @abc.abstractmethod
    def initial_route(self) -> tuple[Route, float]:
        ...
        
    @abc.abstractmethod
    def sample(self, route: Route) -> tuple[Route, float]:
        ...
        
    @abc.abstractmethod
    def acceptance_probability(self, current_val: float, new_val: float, k: int) -> float:
        ...
        
    @abc.abstractmethod
    def stop_criteria(self, stop_counter: int) -> bool:
        ...
        
    @property
    @abc.abstractmethod
    def n_iter(self) -> int:
        ...
        
    @property
    @abc.abstractmethod
    def rng(self) -> Generator:
        ...
        
    def on_step(self, k: int, route: Route, val: float, log_accept_prob: float):
        pass
        
    def on_accept(self, k: int, route: Route, val: float):
        pass

    def run(self) -> tuple[Route, float]:
        route, route_val = self.initial_route()
        
        best_route = route
        best_val = route_val
        
        stop_counter = 0
        current_route = route
        current_val = route_val
        
        for k in range(1, self.n_iter + 1):
            new_route, new_val = self.sample(current_route)
            
            log_accept_prob = self.acceptance_probability(current_val, new_val, k)
            
            if log_accept_prob >= 0 or self.rng.uniform() < np.exp(log_accept_prob):
                # Accepted
                stop_counter = 0
                
                if new_val < best_val:
                    best_route = new_route
                    best_val = new_val
                    
                current_route = new_route
                current_val = new_val
                
                self.on_accept(k, current_route, current_val)
            else:
                stop_counter += 1
                
            self.on_step(k, current_route, current_val, log_accept_prob)
            
            if self.stop_criteria(stop_counter):
                break
                
        return best_route, best_val


@dataclass
class SimulatedAnnealingTSP(AbstractSimulatedAnnealing):
    """
    The Simulated Annealing algorithm for solving the Traveling Salesman Problem.
    Pure Python implementation.
    """
    cities: Cities
    kernel: Kernel
    n_iter: int = 1_000_000
    temperature: cooling_schedule_function = field(default_factory=exponential_cooling_schedule)
    early_stop: bool = True
    stop_after: int = 10_000
    seed: t.Optional[int] = None

    def __post_init__(self):
        if self.seed is not None:
             self._rng = np.random.default_rng(self.seed)
        else:
             if hasattr(self.kernel, "rng"):
                 self._rng = self.kernel.rng
             else:
                 self._rng = np.random.default_rng()

    @property
    def rng(self) -> Generator:
        return self._rng

    def initial_route(self) -> tuple[Route, float]:
        route = Route(list(range(len(self.cities))))
        return route, self.cities.total_distance(route)

    def sample(self, route: Route) -> tuple[Route, float]:
        new_route = self.kernel.sample(route)
        return new_route, self.cities.total_distance(new_route)

    def acceptance_probability(self, current_val: float, new_val: float, k: int) -> float:
        gap = new_val - current_val
        if gap <= 0:
            return 0.0
        return -gap / (self.temperature(k) + _EPS)

    def stop_criteria(self, stop_counter: int) -> bool:
        return self.early_stop and stop_counter > self.stop_after


@dataclass
# SAStatsProxy follows SAStatsProtocol implicitly via fields
class SAStatsProxy(AbstractSimulatedAnnealing):
    """
    Proxy wrapper that collects statistics during SA execution.
    """
    sa: SimulatedAnnealingTSP
    
    # Public fields for statistics
    log_acceptance_probs: list[float] = field(default_factory=list, init=False)
    n_accept: int = field(default=0, init=False)
    acceptance_ratio: list[float] = field(default_factory=list, init=False)
    values: list[float] = field(default_factory=list, init=False)
    # accept_prob protocol requirement: Sequence[float]. 
    # We can compute it on the fly or store it. Keeping it as property might be cleaner if user allows,
    # but request said "campos directos". I'll add a field.
    accept_prob: list[float] = field(default_factory=list, init=False)

    def __post_init__(self):
        # Clear/Init stats
        self.log_acceptance_probs = []
        self.n_accept = 0
        self.acceptance_ratio = []
        self.values = []
        self.accept_prob = []

    # Delegates
    @property
    def n_iter(self) -> int:
        return self.sa.n_iter
        
    @property
    def rng(self) -> Generator:
        return self.sa.rng
        
    def initial_route(self) -> tuple[Route, float]:
        return self.sa.initial_route()
        
    def sample(self, route: Route) -> tuple[Route, float]:
        return self.sa.sample(route)
        
    def acceptance_probability(self, current_val: float, new_val: float, k: int) -> float:
        return self.sa.acceptance_probability(current_val, new_val, k)
        
    def stop_criteria(self, stop_counter: int) -> bool:
        return self.sa.stop_criteria(stop_counter)

    # Hooks for stats
    def on_accept(self, k: int, route: Route, val: float):
        self.n_accept += 1

    def on_step(self, k: int, route: Route, val: float, log_accept_prob: float):
        self.log_acceptance_probs.append(log_accept_prob)
        self.values.append(val)
        self.acceptance_ratio.append(self.n_accept / k)
        self.accept_prob.append(np.exp(log_accept_prob))
        
    def run(self) -> tuple[Route, float]:
        # Reset stats before run
        self.__post_init__()
        return super().run()
